File: v2/src/main/java/com/api/v2/customers/controllers/CustomerController.java
package com.api.v2.customers.controllers;

import com.api.v2.customers.dtos.CustomerRegistrationDto;
import com.api.v2.customers.dtos.CustomerResponseDto;
import com.api.v2.customers.services.CustomerRegistrationService;
import com.api.v2.customers.services.CustomerRetrievalService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("api/v2/customers")
public class CustomerController {

    private final CustomerRegistrationService registrationService;
    private final CustomerRetrievalService retrievalService;

    public CustomerController(
            CustomerRegistrationService registrationService,
            CustomerRetrievalService retrievalService
    ) {
        this.registrationService = registrationService;
        this.retrievalService = retrievalService;
    }

    @Operation(summary = "Register a customer")
    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    public Mono<CustomerResponseDto> register(@Valid @RequestBody CustomerRegistrationDto registrationDto) {
        return registrationService.register(registrationDto);
    }

    @GetMapping("{id}")
    @Operation(summary = "Retrieve a customer by its id")
    public Mono<CustomerResponseDto> findById(@PathVariable String id) {
        return retrievalService.findById(id);
    }

    @Operation(summary = "Retrieve all customers")
    @GetMapping
    public Flux<CustomerResponseDto> findAll() {
        return retrievalService.findAll();
    }
}

File: v2/src/main/kotlin/com/api/v2/customers/controllers/CustomerController.kt
package com.api.v2.customers.controllers

import com.api.v2.customers.domain.Customer
import com.api.v2.customers.dtos.CustomerModificationDto
import com.api.v2.customers.dtos.CustomerRegistrationDto
import com.api.v2.customers.services.CustomerModificationService
import com.api.v2.customers.services.CustomerRegistrationService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v2/customers")
class CustomerController(
    private val registrationService: CustomerRegistrationService,
    private val modificationService: CustomerModificationService
) {

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    suspend fun register(@RequestBody registrationDto: @Valid CustomerRegistrationDto): Customer {
        return registrationService.register(registrationDto)
    }

    @PutMapping("{ssn}")
    suspend fun modify(
        @PathVariable ssn: String,
        @RequestBody customerModificationDto: @Valid CustomerModificationDto
    ) {
        return modificationService.modify(ssn, customerModificationDto)
    }

}